### PR TITLE
ARROW-2278: [Python] Fix incorrect numpy struct array serialization

### PR DESCRIPTION
**Pull request by Licht-T**
This fixes [ARROW-2278](https://issues.apache.org/jira/browse/ARROW-2278).

### CODE REVIEW

File: python/pyarrow/serialization.py [hunk 1]
```diff
@@ -30,27 +31,83 @@
...
 def _serialize_numpy_array_list(obj):
     if obj.dtype.str != '|O':
         # Make the array c_contiguous if necessary so that we can call change
         # the view.
         if not obj.flags.c_contiguous:
             obj = np.ascontiguousarray(obj)
-        return obj.view('uint8'), obj.dtype.str
+
+        if obj.dtype.fields is not None:
+            dtype = DTypeJsonEncoder().encode(dict(obj.dtype.fields))
```
Comment (pitrou) [inline python/pyarrow/serialization.py]: I think it's overkill to bring json into this and customize the encoder. You could simply use pickle, which Numpy dtypes support natively (you could also use `dtype.descr`, but it won't work with all dtypes).

File: python/pyarrow/tests/test_serialization.py [hunk 2]
```diff
@@ -27,6 +27,7 @@
 
 import pyarrow as pa
 import numpy as np
+import pandas.util.testing as tm
```
Comment (pitrou) [inline python/pyarrow/tests/test_serialization.py]: You don't need this, just use `numpy.testing.assert_array_equal`.

**Review by pitrou — CHANGES_REQUESTED** [review]
@Licht-T, thanks for posting this. Please address the review comments below.

**Comment by Licht-T** [comment]
Thanks @pitrou, fixed!

**Comment by pitrou** [comment]
Yes, `npm` looks flaky these days unfortunately.

File: python/pyarrow/serialization.py [hunk 3]
```diff
@@ -42,15 +43,24 @@ def _serialize_numpy_array_list(obj):
...
 
 
 def _deserialize_numpy_array_list(data):
     if data[1] != '|O':
         assert data[0].dtype == np.uint8
-        return data[0].view(data[1])
+
+        try:
+            dtype = pickle.loads(data[1])
+        except (TypeError, pickle.UnpicklingError):
```
Comment (pitrou) [inline python/pyarrow/serialization.py]: I don't think we should try to unpickle data that we know isn't a pickle (in extreme cases there may even be false positives, though that's unlikely). Rather, perhaps we should change the serialization format for Numpy arrays. For example a 3-tuple `(kind, serialized dtype, data)`. The `kind` (for example a string: "object", "struct", "scalar") would help discriminate between different serialization formats. After all we already have a distinct deserialization path for object arrays.

**Comment by wesm** [comment]
@Licht-T I know it's been a long time but would like to rebase and perhaps address @pitrou's feedback?

**Comment by wesm** [comment]
Closing for now for staleness

**Pull request closed without merging**